src/core/Sample.cpp

```diff
@@ -0,0 +1,414 @@
+	swap(first.m_loopEndFrame, second.m_loopEndFrame);
+	swap(first.m_amplification, second.m_amplification);
+	swap(first.m_frequency, second.m_frequency);
+	swap(first.m_reversed, second.m_reversed);
+}
+
+bool Sample::play(sampleFrame* dst, PlaybackState* state, int numFrames, float desiredFrequency, Loop loopMode) const
+{
+	auto lock = std::shared_lock{m_mutex};
+	if (numFrames <= 0 || desiredFrequency <= 0) { return false; }
+
+	auto resampleRatio = static_cast<float>(Engine::audioEngine()->processingSampleRate()) / m_buffer->sampleRate();
+	resampleRatio *= m_frequency / desiredFrequency;
+
+	auto playBuffer = std::vector<sampleFrame>(numFrames / resampleRatio);
+	if (!typeInfo<float>::isEqual(resampleRatio, 1.0f))
+	{
+		playBuffer.resize(playBuffer.size() + s_interpolationMargins[state->m_interpolationMode]);
+	}
+
+	switch (loopMode)
+	{
+	case Loop::Off:
+		state->m_frameIndex = std::clamp(state->m_frameIndex, m_startFrame, m_endFrame);
+		if (state->m_frameIndex == m_endFrame) { return false; }
+		break;
+	case Loop::On:
+		state->m_frameIndex = std::clamp(state->m_frameIndex, m_startFrame, m_loopEndFrame);
+		if (state->m_frameIndex == m_loopEndFrame) { state->m_frameIndex = m_loopStartFrame; }
+		break;
+	case Loop::PingPong:
+		state->m_frameIndex = std::clamp(state->m_frameIndex, m_startFrame, m_loopEndFrame);
+		if (state->m_frameIndex == m_loopEndFrame)
+		{
+			state->m_frameIndex = m_loopEndFrame - 1;
+			state->m_backwards = true;
+		}
+		else if (state->m_frameIndex <= m_loopStartFrame && state->m_backwards)
+		{
+			state->m_frameIndex = m_loopStartFrame;
+			state->m_backwards = false;
+		}
+		break;
+	}
+
+	playSampleRange(state, playBuffer.data(), playBuffer.size());
+
+	auto resample = resampleSampleRange(
+		state->m_resampleState, playBuffer.data(), dst, playBuffer.size(), numFrames, resampleRatio);
+	state->m_frameIndex += (state->m_backwards ? -1 : 1) * resample.input_frames_used;
+	if (src_error(state->m_resampleState) != 0) { return false; }
+
+	amplifySampleRange(dst, numFrames);
+	return true;
+}
+
+void Sample::visualize(QPainter& p, const QRect& dr, int fromFrame, int toFrame) const
+{
+	const auto lock = std::shared_lock{m_mutex};
+
+	const auto x = dr.x();
+	const auto height = dr.height();
+	const auto width = dr.width();
+	const auto centerY = dr.center().y();
+
+	const auto halfHeight = height / 2;
+	const auto buffer = m_buffer->data() + fromFrame;
+
+	const auto color = p.pen().color();
+	const auto rmsColor = color.lighter(123);
+
+	auto numFrames = toFrame - fromFrame;
+	if (numFrames == 0) { numFrames = m_buffer->size(); }
+
+	const float framesPerPixel = numFrames / static_cast<float>(width);
+
+	constexpr auto minResolution = 512;
+	auto resolution = 1;
+	while (framesPerPixel / resolution > minResolution)
+	{
+		resolution++;
+	}
+
+	auto min = std::vector<float>(width);
+	auto max = std::vector<float>(width);
+	auto rms = std::vector<float>(width);
```

Comment:
```suggestion
	auto min = std::vector<float>(width, 1);
	auto max = std::vector<float>(width, -1);
	auto rms = std::vector<float>(width, 0);
```

These vectors should be initialized like this, else min and max won't update correctly with some values above/below 0. Here is an example:
![image](https://github.com/LMMS/lmms/assets/47889291/d2b2c4ed-3b2a-414e-a4b2-d4a9c1b6aaa5)
`rms` should also initialize with 0 to avoid starting with garbage data.

Reply:
I don't fully understand why the `min` and `max` values have to be set to that. What exactly do you mean by "some values"? What kind of values are causing a problem? Though, I think it could be easier to just remove the allocations and simply choose the largest sample value, which mimics how Audacity does it.

Also, `rms` shouldn't even have to be set to 0 explicitly because `std::vector`'s constructor will default insert the types, which in this will do zero initialization since they are floats.

Reply:
> I don't fully understand why the min and max values have to be set to that. What exactly do you mean by "some values"? What kind of values are causing a problem?

Let's take min as an example. If it is set to 0, all values that are above 0 get disregarded, since that value is going to be bigger than 0. The same applies for max but with values smaller than 0. The result is that either min or max will stay at 0, as you can see in the example. 

> Though, I think it could be easier to just remove the allocations and simply choose the largest sample value, which mimics how Audacity does it.

I don't exactly understand, but that sounds exactly like what is being done. The vectors are only there to have the least amount of operations inside that large for loop.

> Also, rms shouldn't even have to be set to 0 explicitly because std::vector's constructor will default insert the types, which in this will do zero initialization since they are floats.

Yeah sorry I didn't know that, in that case there is no need to initialize `rms`. Although I do think it's clearer for people that don't know.

Reply:
> Let's take min as an example. If it is set to 0, all values that are above 0 get disregarded, since that value is going to be bigger than 0. The same applies for max but with values smaller than 0. The result is that either min or max will stay at 0, as you can see in the example.

Ah, makes sense, which is probably why it is better to determine the most minimum and most maximum sample values from the range of samples themselves, and not refer back to a vector of floats where its initialization can mess with things.

> I don't exactly understand, but that sounds exactly like what is being done. The vectors are only there to have the least amount of operations inside that large for loop.

What I mean is to find the sample with the highest amplitude, either negative or positive, and draw the line based on that alone (so no need to find the minimum).

Reply:
> Ah, makes sense, which is probably why it is better to determine the most minimum and most maximum sample values from the range of samples themselves, and not refer back to a vector of floats where its initialization can mess with things.

Even if you remove the vectors, you would still have to initialize a min and max variable as +1 and -1 respectively. The vector is just to keep the main processing loop and the drawing loop separate for performance reasons. Sorry if I misunderstood what you said.

> What I mean is to find the sample with the highest amplitude, either negative or positive, and draw the line based on that alone (so no need to find the minimum).

Doesn't work, since if you had a sample that for example was a straight line at 0.5, it would draw a rectangle from -0.5 to 0.5, since it doesn't have a minimum to work with. rms would also break.

Unrelated, but you should probably add a check to see if the sample is empty to not render anything if it is.

Reply:
> Surely this works? Sorry if I misunderstood your concern.

Maybe I expressed myself wrong, but I'm pretty confident it doesn't work. I initially thought the same as you, but once I implement it, I saw that it doesn't work with short samples. There is in fact a reason that the original function did it like that.

Look at this sample, if you only use the max, how do you draw the line? 

![image](https://github.com/LMMS/lmms/assets/47889291/492cc9ef-eae4-4735-899b-434c6681eaf0)

You can't draw from 0 to max, since it would look like my initial example. 
You also can't draw from -max to max, since you would end up with something perfectly symmetrical. 
The only option is to draw from min to max, and since they are close together, the line looks like a dot.

Of course it would be great if you could simplify even more, but I already tried quite a few things to get to this point.

Reply:
Yeah, you're right, we have to get the most negative and most positive sample value.  I still want to remove those vector allocations though.

Reply:
Maybe somewhat complicated, but you could probably combine everything into one loop, something like this:
```C++
// min, max, rms init here
int segmentIndex = 0;
for (int i = 0; i < frames - resolution; i += resolution) // same main loop
{
	pixelIndex = i / framesPerPixel;
	// update min max rms
	if (pixelIndex > segmentIndex) // everytime pixelIndex increases by integer amount
	{
		// draw here using segmentIndex as index
		// reset min max rms
		segementIndex++;
	}
}
```

I don't really see a reason to remove the allocations, but it's definitly possible. I also think that this pseudocode avoids a bug with `framesPerPixel` being less than 1. Whatever you do will be faster than the original implementation anyway, so it shouldn't matter that much.

Reply:
> For some reason, I was thinking it was allocating them for every single frame, not just the width of the box 🤦 it should also be very cheap since its just float's.

Oh well, no problem then.

> I think everything is fine then after I apply the original suggestion you made, but you just mentioned that there's the bug for if framesPerPixel is < 1. I think for that to happen, we didn't have enough frames to fill in the entire box. I'm thinking we could just set the minimum to 1 then?

Problem is that the vectors don't get changed and retain their original values. Easiest fix I can think of is setting min to `something > 1`, and then in the drawing loop check for that number and skip if necessary. Alternatively you could calculate the actual pixels used and create the vector with that, and later scale to the box. Sorry, I don't really have time to write an example right now, but it should be a simple change.